File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/usecases/TopUpRequesting.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.common.application.events.EventPublisher;
import it.giannibombelli.wsc2026.giftcard.application.commands.RequestGiftCardTopUp;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardEvent;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardTopUpRequested;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;

import static java.util.Objects.requireNonNull;

public final class TopUpRequesting implements UseCase<RequestGiftCardTopUp, GiftCardTopUpRequested> {
    private final GiftCardRepository giftCardRepository;
    private final EventPublisher<GiftCardEvent> eventPublisher;

    public TopUpRequesting(GiftCardRepository giftCardRepository, EventPublisher<GiftCardEvent> eventPublisher) {
        this.giftCardRepository = requireNonNull(giftCardRepository);
        this.eventPublisher = requireNonNull(eventPublisher);
    }

    @Override
    public GiftCardTopUpRequested invoke(RequestGiftCardTopUp cmd) {
        Require.requireArgument(cmd, "command");

        final GiftCardId giftCardId = cmd.aggregateId();
        GiftCard giftCard = giftCardRepository.findById(giftCardId)
            .orElseThrow(() -> new IllegalStateException("Gift card not found: " + giftCardId.value()));

        GiftCardTopUpRequested topUpRequested = giftCard.requestTopUp(cmd.amount());

        giftCardRepository.save(giftCard);
        eventPublisher.publish(topUpRequested);

        return topUpRequested;
    }
}

File: typescript/src/giftcard/application/usecases/topUpRequesting.ts
// Caso d'uso per la richiesta di ricarica di una gift card.

import type { EventPublisher } from '@/common/application/events/eventPublisher.js';
import type { UseCase } from '@/common/application/usecase.js';
import { GiftCardEvent } from '../../domain/events/giftCardEvent.js';
import { GiftCardTopUpRequested } from '../../domain/events/giftCardTopUpRequested.js';
import { GiftCardId } from '../../domain/giftcard/giftCardId.js';
import type { GiftCardRepository } from '../../domain/ports/giftCardRepository.js';
import { RequestGiftCardTopUp } from '../commands/requestGiftCardTopUp.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class TopUpRequesting implements UseCase<RequestGiftCardTopUp, GiftCardTopUpRequested> {
  private readonly _giftCardRepository: GiftCardRepository;
  private readonly _eventPublisher: EventPublisher<GiftCardEvent>;

  constructor(giftCardRepository: GiftCardRepository, eventPublisher: EventPublisher<GiftCardEvent>) {
    requireDependency(giftCardRepository, "giftCardRepository");
    requireDependency(eventPublisher, "eventPublisher");
    this._giftCardRepository = giftCardRepository;
    this._eventPublisher = eventPublisher;
  }

  invoke(command: RequestGiftCardTopUp): GiftCardTopUpRequested {
    requireArgument(command, 'command');

    const giftCardId: GiftCardId = command.aggregateId;
    const giftCard = this._giftCardRepository.findById(giftCardId);
    if (giftCard === null) {
      throw new Error(`Gift card not found: ${giftCardId.value}`);
    }

    const topUpRequested = giftCard.requestTopUp(command.amount);

    this._giftCardRepository.save(giftCard);
    this._eventPublisher.publish(topUpRequested);

    return topUpRequested;
  }
}
